pkg/suggestion/v1alpha3/hyperopt_service.py

```diff
@@ -40,17 +41,53 @@ def GetSuggestions(self, request, context):
         )
 
     def ValidateAlgorithmSettings(self, request, context):
+        is_valid, message = OptimizerConfiguration.validate_algorithm_spec(
+            request.experiment.spec.algorithm)
+        if not is_valid:
+            context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+            context.set_details(message)
+            logger.error(message)
         return api_pb2.ValidateAlgorithmSettingsReply()
 
 
-class OptimizerConfiguration(object):
-    def __init__(self, random_state=None):
-        self.random_state = random_state
+class OptimizerConfiguration:
+    __schema_dict = {
+        'tpe': {
+            'gamma': (lambda x: float(x), lambda x: 1 > float(x) > 0),
+            'prior_weight': (lambda x: float(x), lambda x: float(x) > 0),
+            'n_EI_candidates': (lambda x: int(x),
+                                lambda x: float(x).is_integer() and int(x) > 0),
+            "random_state": (lambda x: int(x), lambda x: x.isdigit()),
+        },
+        "random": {
+            "random_state": (lambda x: int(x), lambda x: x.isdigit())
+        }
+    }
```

Comment:
I mind that validate functions ( `__schema_dict[algo_name][s.name][1]` ) cannot return the message why parameters are invalid.

Reply:
And I personally feels this dict is a bit tricky. I prefer the following code:

```python
@classmethod
def _validate_tpe_setting(algorithm_settings):
    for s in algorithm_settings:
        if s == 'gamma' and 0 < float(x) < 1:
            return False, "gamma should be in the range of [0, 1]"
        ...
    return True, ""

@classmethod
def validate_algorithm_spec(cls, algorithm_spec):
    algo_name = algorithm_spec.algorithm_name
    if algo_name == "tpe":
        return cls._validate_tpe_setting(algorithm_spec.algorithm_setting)
    elif algo_name == "random":
        return cls._validate_random_setting(algorithm_spec.algorithm_setting)
    return False, "invalid algorithm name"
```

But I want to hear opinions of maintainers ( @andreyvelich @gaocegege ).

Reply:
I'm okay with both code styles.

Reply:
I agree with @c-bata, I think for validation we should say why settings is not correct. Like we did with enas: https://github.com/kubeflow/katib/blob/master/pkg/suggestion/v1alpha3/nas/enas_service.py#L218

Reply:
@andreyvelich With current implementaion, we also tell users why settings is wrong: https://github.com/kubeflow/katib/pull/1161/files#diff-a9d0a87b0b01a9816c0583cd6a0003e2R244

Reply:
@c-bata @andreyvelich I've refactored codes of validation part.